lib/Fhp/Action/SendSEPADirectDebit.php

```diff
@@ -3,11 +3,100 @@
 namespace Fhp\Action;
 
 use Fhp\BaseAction;
+use Fhp\DataTypes\Bin;
+use Fhp\Model\SEPAAccount;
+use Fhp\Segment\BaseSegment;
+use Fhp\Segment\Common\Btg;
+use Fhp\Segment\Common\Kti;
+use Fhp\Segment\DME\HIDMESv1;
+use Fhp\Segment\DME\HIDMESv2;
+use Fhp\Segment\DME\HIDXES;
+use Fhp\Segment\DME\HKDMEv1;
+use Fhp\Segment\DSE\HIDSESv2;
+use Fhp\Segment\DSE\HKDSEv1;
+use Fhp\Segment\SPA\HISPAS;
+use Fhp\UnsupportedException;
 
 class SendSEPADirectDebit extends BaseAction
 {
+    /** @var SEPAAccount */
+    private $account;
+
+    /** @var string */
+    private $painMessage;
+
+    /** @var string */
+    private $xmlSchema;
+
+    /** @var float */
+    private $ctrlSum;
+
+    /** @var bool */
+    private $singleDirectDebit = false;
+
+    public static function create(SEPAAccount $account, string $painMessage): SendSEPADirectDebit
+    {
+        if (preg_match('/xmlns="(?<namespace>[^"]+)".*<GrpHdr>.*<CtrlSum>(?<ctrlsum>[.0-9]+)<\/CtrlSum>.*<\/GrpHdr>/s', $painMessage, $matches) === 1) {
+            $painUrn = $matches['namespace'];
+            $ctrlSum = $matches['ctrlsum'];
+        } else {
+            throw new \InvalidArgumentException('PAIN-Namespace and or <GrpHdr><CtrlSum>xx</CtrlSum></GrpHdr> missing in PAIN message');
+        }
+
+        // Check wether it is one DirectDebit or multiple, should match <NbOfTxs>xx</NbOfTxs> in the XML
+        $nbOfTxs = substr_count($painMessage, '<DrctDbtTxInf>');
+
+        $result = new SendSEPADirectDebit();
+        $result->account = $account;
+        $result->painMessage = $painMessage;
+        $result->xmlSchema = $painUrn;
+        $result->ctrlSum = $ctrlSum;
+        $result->singleDirectDebit = $nbOfTxs == 1;
+        return $result;
+    }
+
     public function createRequest($bpd, $upd)
     {
-        return [];
+        $type = $this->singleDirectDebit ? 'S' : 'M';
+
+        /** @var HIDXES|BaseSegment $hidxes */
+        $hidxes = $bpd->requireLatestSupportedParameters('HID' . $type . 'ES');
```

Comment:
Ich würde einfach hier `$this->singleDirectDebit ? 'HIDSES' : 'HIDMES'` machen, und unten so ähnlich, sodass es die `$type`-Variable nicht braucht. Das braucht ein bisschen mehr Programmcode, wird aber schneller ausgeführt (keine Stringkonkatenation) und hat vor allem den Vorteil, dass man global im Projekt nach "HIDMES" suchen kann und dann diese Stelle hier findet.

Reply:
Aus ähnlichen Gründen ist bspw. [hier](https://github.com/nemiah/phpFinTS/blob/ef25fafe7dc09ee4dbb4d5f98229ab4b8c51566a/lib/Fhp/Action/GetSEPAAccounts.php#L51) ein `switch` und kein Klassenname der aus einem String hergeleitet wird. Dann kann auch die IDE die Verknüpfung herstellen und es sieht nicht so aus, als würde niemand das Segment verwenden (und man könnte es löschen). Würde ich unten auch so machen, anstatt den Klasennamen zusammenzubasteln.